packages/sources/intrinio/src/util.ts

```diff
@@ -0,0 +1,296 @@
+
+const EventEmitter = events.EventEmitter
+
+const SELF_HEAL_BACKOFFS = [0, 100, 500, 1000, 2000]
+
+export class IntrinioRealtime extends EventEmitter {
+  options: any
+  token: null | string = null
+  websocket = null
+  ready = false
+  channels = {}
+  joinedChannels = {}
+  afterConnected = null // Promise
+  self_heal_backoff = Array.from(SELF_HEAL_BACKOFFS)
+  self_heal_ref = null
+  quote_callback = null
+  error_callback = null
+
+  constructor(options: any) {
+    super()
+
+    this.options = options
+    this.token = null
+    this.websocket = null
+    this.ready = false
+    this.channels = {}
+    this.joinedChannels = {}
+    this.afterConnected = null // Promise
+    this.self_heal_backoff = Array.from(SELF_HEAL_BACKOFFS)
+    this.self_heal_ref = null
+    this.quote_callback = null
+    this.error_callback = null
+
+    // Parse options
+    if (!options) {
+      this._throw('Need a valid options parameter')
+    }
+
+    if (options.api_key) {
+      if (!this._validAPIKey(options.api_key)) {
+        this._throw('API Key was formatted invalidly')
+      }
+    } else {
+      if (!options.username && !options.password) {
+        this._throw('API key or username and password are required')
+      }
+
+      if (!options.username) {
+        this._throw('Need a valid username')
+      }
+
+      if (!options.password) {
+        this._throw('Need a valid password')
+      }
+    }
+
+    const providers = ['iex', 'quodd', 'cryptoquote', 'fxcm']
+    if (!options.provider || !providers.includes(options.provider)) {
+      this._throw('Need a valid provider: iex, quodd, cryptoquote, or fxcm')
+    }
+  }
+
+  _throw(e: any) {
+    let handled = false
+    if (typeof e === 'string') {
+      e = 'IntrinioRealtime | ' + e
+    }
+    if (this.listenerCount('error') > 0) {
+      Logger.error(e)
+      handled = true
+    }
+    if (!handled) {
+      throw e
+    }
+  }
+
+  _makeAuthUrl(): { host: string; path: string } {
+    let auth_url = {
+      host: '',
+      path: '',
+    }
+
+    if (this.options.provider == 'iex') {
+      auth_url = {
+        host: 'realtime.intrinio.com',
+        path: '/auth',
+      }
+    } else if (this.options.provider == 'quodd') {
+      auth_url = {
+        host: 'api.intrinio.com',
+        path: '/token?type=QUODD',
+      }
+    } else if (this.options.provider == 'cryptoquote') {
+      auth_url = {
+        host: 'crypto.intrinio.com',
+        path: '/auth',
+      }
+    } else if (this.options.provider == 'fxcm') {
+      auth_url = {
+        host: 'fxcm.intrinio.com',
+        path: '/auth',
+      }
+    }
+
+    if (this.options.api_key) {
+      auth_url = this._makeAPIAuthUrl(auth_url)
+    }
+
+    return auth_url
+  }
+
+  _makeAPIAuthUrl(auth_url: any): any {
+    let path = auth_url.path
+
+    if (path.includes('?')) {
+      path = path + '&'
+    } else {
+      path = path + '?'
+    }
+
+    auth_url.path = path + 'api_key=' + this.options.api_key
```

Comment:
```suggestion
    return auth_url.path = `${auth_url.path}${auth_url.path.includes('?') ? '&' : '?'}api_key=${this.options.api_key}` 
```
How do you feel about making this more terse?

Reply:
BTW, we may want to use the URL node lib here instead, so we dont have to manually manipulate strings